ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -0,0 +1,234 @@
+   * @param containerValidSizeMap the valid data size per container map
+   * @param referenceTimeInMs the reference time used to decide whether or not a record is valid
+   * @param messageInfo the {@link MessageInfo} to be processed
+   * @param backwardSearchSpan a {@link FileSpan} that spans from the beginning of the index to the beginning of the
+   *                           previous index segment and is used for looking up corresponding put record when applicable
+   * @throws StoreException
+   */
+  private void populateContainerMap(Map<String, Map<String, Long>> containerValidSizeMap, long referenceTimeInMs,
+      MessageInfo messageInfo, FileSpan backwardSearchSpan) throws StoreException {
+
+    if (messageInfo.isDeleted()) {
+      // delete record
+      IndexValue putIndexValue = index.findKey(messageInfo.getStoreKey(), backwardSearchSpan);
+      if (putIndexValue != null && !putIndexValue.isFlagSet(IndexValue.Flags.Delete_Index) && !isExpired(
+          putIndexValue.getExpiresAtMs(), referenceTimeInMs)) {
+        // correct previously counted value only if the following conditions are true:
+        // 1. the put and delete record are not in the same index segment (putIndexValue != null)
+        // 2. the found record is actually a put record (!putIndexValue.isFlagSet)
+        // 3. the found put record is not expired
+        updateNestedMapHelper(containerValidSizeMap, BlobStoreStats.getServiceId(messageInfo.getStoreKey()),
+            BlobStoreStats.getContainerId(messageInfo.getStoreKey()), putIndexValue.getSize() * -1);
+      }
+    } else if (!isExpired(messageInfo.getExpirationTimeInMs(), referenceTimeInMs)) {
+      // put record that is not expired
+      updateNestedMapHelper(containerValidSizeMap, BlobStoreStats.getServiceId(messageInfo.getStoreKey()),
+          BlobStoreStats.getContainerId(messageInfo.getStoreKey()), messageInfo.getSize());
+    }
+  }
+
+  /**
+   * Process a record in the index and populate the valid data size per log segment map.
+   * @param logSegmentValidSizeMap the valid data size per log segment map
+   * @param referenceTimeInMs the reference time used to decide whether or not a record is valid
+   * @param messageInfo the {@link MessageInfo} to be processed
+   * @param lastModifiedTimeInMs the time stamp indicating when the record that is being processed was added
+   * @param logSegmentName the log segment name of the record that is being processed
+   * @param backwardSearchSpan a {@link FileSpan} that spans from the beginning of the index to the beginning of the
+   *                           previous index segment and is used for looking up corresponding put record when applicable
+   * @throws StoreException
+   */
+  private void populateLogSegmentMap(Map<String, Long> logSegmentValidSizeMap, long referenceTimeInMs,
+      MessageInfo messageInfo, long lastModifiedTimeInMs, String logSegmentName, FileSpan backwardSearchSpan)
+      throws StoreException {
+
```

Comment:
Once #541 is merged, one option to remove the duplicate code here is to process a list of `IndexValue` and return a list of valid `IndexValue` that needs to be counted. For example, a valid put will return its `IndexValue` and depending on the caller, the caller can either update the log segment valid size map or the container valid size map using the returned information. This change is not possible at the moment because existing IndexValue does not contain serviceId and containerId yet.